### PR TITLE
[docs] clarify the raised exception

### PR DESCRIPTION
**Pull request by ringabout**
Lest developers wanna know what the exception is.

### CODE REVIEW

File: lib/pure/uri.nim [hunk 1]
```diff
@@ -152,9 +152,9 @@ func encodeQuery*(query: openArray[(string, string)], usePlus = true,
 
 iterator decodeQuery*(data: string): tuple[key, value: string] =
   ## Reads and decodes query string `data` and yields the `(key, value)` pairs
-  ## the data consists of. If compiled with `-d:nimLegacyParseQueryStrict`, an
-  ## error is raised when there is an unencoded `=` character in a decoded
-  ## value, which was the behavior in Nim < 1.5.1
+  ## the data consists of. If compiled with `-d:nimLegacyParseQueryStrict`,
+  ## `UriParseError` is raised when there is an unencoded `=` character in a decoded
+  ## value, which was the behavior in Nim < 1.5.1.
```
Comment (xflywind) [inline lib/pure/uri.nim:157]:
```suggestion
  ## value, which was the behavior in Nim < 1.5.1. Otherwise it doesn't raise.
```

**Comment by ringabout** [comment]
https://github.com/nim-lang/Nim/pull/16647/files is a breaking change. It should mention `cgi.decodeData` doesn't raise `CgiError` anymore without `-d:nimLegacyParseQueryStrict` in docs.

For example:
https://github.com/dom96/jester/blob/471b4b8d3c4f40278a5b0fb791be836f98ca4ed0/jester/request.nim#L96 still catch `CgiError` which doesn't make sense anymore.

And even with `d:nimLegacyParseQueryStrict`, it never raises `CgiError` now, which can break users' code potentially.

**Comment by arnetheduck** [comment]
> is a breaking change

the easy way to notice these _and_ document the API is to annotate the function explicitly with `{.raises: [UriParseError].}` - this way we avoid breaking changes like this in the future.

**Comment by arnetheduck** [comment]
that said, _not_ raising an exception any more is a benign change: nothing that worked before breaks, generally - raising a different or new exception is a breaking change - nonetheless, the explicit annotation ensures that the _public_ API remains stable.

**Comment by ringabout** [comment]
> that said, not raising an exception any more is a benign change: nothing that worked before breaks, generally

I agree. And the docs of `cgi.decodeData` should clearly clarify them, so that the library author can probably refactor their code in the future.

For example, 

```nim
  try:
    for key, val in cgi.decodeData(query):
      result[key] = val
  except CgiError:
    logging.warn("Incorrect query. Got: $1" % [query]) 
```
can become
```nim
    for key, val in cgi.decodeData(query):
      result[key] = val
```

eventually and reduce the cost to the minimal (try except is not zero cost in Nim).

**Comment by ringabout** [comment]
> that said, not raising an exception any more is a benign change: nothing that worked before breaks, generally - raising a different or new exception is a breaking change 

It can cause troubles in one way or another. If users call `cgi.decodeData` with `-d:nimLegacyParseQueryStrict` option, it breaks code. Because before it raises `CgiError` and now it raises `UriParseError`.

File: lib/pure/uri.nim [hunk 2]
```diff
@@ -152,9 +152,9 @@ func encodeQuery*(query: openArray[(string, string)], usePlus = true,
 
 iterator decodeQuery*(data: string): tuple[key, value: string] =
   ## Reads and decodes query string `data` and yields the `(key, value)` pairs
```
Comment (konsumlamm) [inline lib/pure/uri.nim]:
```suggestion
  ## Reads and decodes the query string `data` and yields the `(key, value)` pairs
```

File: lib/pure/uri.nim [hunk 3]
```diff
@@ -152,9 +152,9 @@ func encodeQuery*(query: openArray[(string, string)], usePlus = true,
...
-  ## the data consists of. If compiled with `-d:nimLegacyParseQueryStrict`, an
-  ## error is raised when there is an unencoded `=` character in a decoded
-  ## value, which was the behavior in Nim < 1.5.1
+  ## the data consists of. If compiled with `-d:nimLegacyParseQueryStrict`,
+  ## `UriParseError` is raised when there is an unencoded `=` character in a decoded
```
Comment (konsumlamm) [inline lib/pure/uri.nim]:
```suggestion
  ## a `UriParseError` is raised when there is an unencoded `=` character in a decoded
```

**Pull request merged**